### PR TITLE
Mmc v2

### PR DESCRIPTION
**Pull request by hzhuang1**
Changelog:
v2:
   * Move dwmmc out of this pull request.
   * Rename mmc to emmc.
   * Fix the unused valure issue of release build.

https://github.com/ARM-software/tf-issues/issues/376

### CODE REVIEW

File: drivers/io/io_block.c [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+	io_block_spec_t *region;
+
+	assert((dev_info->info != (uintptr_t)NULL) &&
+	       (spec != (uintptr_t)NULL) &&
+	       (entity->info == (uintptr_t)NULL));
+
+	block = (io_block_entity_t *)spec;
+	region = &block->region;
+	memcpy((void *)&block->buffer, (void *)dev_info->info,
+	       sizeof(io_block_spec_t));
```
Comment (danh-arm) [inline drivers/io/io_block.c]: From this I can infer that you expect platform code to pass an `io_block_spec_t` as the "block device spec" in its call to `io_dev_open()`. But:

*   It's not clear to me what this `io_block_spec_t` represents. Some comments or more info in the commit message would help.
*   It seems unnecessary to copy this into `block->buffer` since it's already stored in the global `block_dev_info.info`. If needed you could provide a macro that casts this to the appropriate type.

File: include/drivers/io/io_block.h [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+typedef struct io_block_entity {
+	io_block_spec_t		region;
+	io_block_spec_t		buffer;
+	io_block_ops_t		ops;
+	size_t			file_pos;
+	int			block_size;
+	int			lba;
+	int			rpmb_enable;
+} io_block_entity_t;
```
Comment (danh-arm) [inline include/drivers/io/io_block.h]: It seems this structure is conflating 2 things; the "image spec" that is passed in the call to `block_open()`, and the internal structure that the driver uses to maintain state. As far as I can see, the "image spec" is entirely represented by the `region` field so you can just define this in platform code instead of an `io_block_entity`.

If you do need to define a new "image spec" type, then this should be defined in `io_storage.h` with the other related types. Hopefully this is not needed since it would be confusing for the "IO block driver" to require an image spec that is different to the existing `io_block_spec_t`.

`io_block_entity` should be private to the driver and declared in the c file or a private header.

File: include/drivers/io/io_block.h [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+typedef struct io_block_ops {
+	size_t	(*read)(int lba, uintptr_t buf, size_t size);
+	size_t	(*write)(int lba, uintptr_t buf, size_t size);
+} io_block_ops_t;
+
+typedef struct io_block_entity {
+	io_block_spec_t		region;
+	io_block_spec_t		buffer;
+	io_block_ops_t		ops;
```
Comment (danh-arm) [inline include/drivers/io/io_block.h]: It looks like `buffer` and `ops` are fields that should be part of the "device spec" passed to `io_dev_open()`. Perhaps you need to define a new `io_block_dev_spec` structure to represent this?

File: include/drivers/io/io_block.h [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+	size_t			file_pos;
+	int			block_size;
+	int			lba;
+	int			rpmb_enable;
```
Comment (danh-arm) [inline include/drivers/io/io_block.h]: rpmb is an eMMC specific thing. Should not be part of this interface.

File: drivers/io/io_block.c [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+#include <assert.h>
+#include <debug.h>
+#include <errno.h>
+#include <io_block.h>
+#include <io_driver.h>
+#include <io_storage.h>
+#include <string.h>
+
+#define MIN_BLOCK_SIZE			64
+#define MIN_BLOCK_MASK			(MIN_BLOCK_SIZE - 1)
```
Comment (danh-arm) [inline drivers/io/io_block.c]: Why is this needed in the generic block driver? Should this check be in the eMMC driver instead?

File: drivers/io/io_block.c [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+		      io_entity_t *entity)
+{
+	io_block_entity_t *block;
+	io_block_spec_t *region;
+
+	assert((dev_info->info != (uintptr_t)NULL) &&
+	       (spec != (uintptr_t)NULL) &&
+	       (entity->info == (uintptr_t)NULL));
+
+	block = (io_block_entity_t *)spec;
```
Comment (danh-arm) [inline drivers/io/io_block.c]: As stated elsewhere, it seems to me that `spec` should only contain `region`.

File: include/drivers/io/io_block.h [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+	size_t	(*write)(int lba, uintptr_t buf, size_t size);
+} io_block_ops_t;
+
+typedef struct io_block_entity {
+	io_block_spec_t		region;
+	io_block_spec_t		buffer;
+	io_block_ops_t		ops;
+	size_t			file_pos;
+	int			block_size;
+	int			lba;
```
Comment (danh-arm) [inline include/drivers/io/io_block.h]: Why do you need to store the `lba` in this structure? It seems that this can be calculated at any time using the other fields in this structure.

File: drivers/io/io_block.c [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+{
+	io_block_entity_t *block;
+	io_block_spec_t *region;
+
+	assert(entity->info != (uintptr_t)NULL);
+
+	block = (io_block_entity_t *)entity->info;
+	region = &block->region;
+
+	assert(offset < block->region.length);
```
Comment (danh-arm) [inline drivers/io/io_block.c]: Perhaps also check offset >=0

File: drivers/io/io_block.c [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+
+	block->file_pos += offset;
+	block->lba = (block->file_pos + region->offset) / block->block_size;
+	assert(block->lba >= 0);
+	return 0;
+}
+
+static int block_read(io_entity_t *entity, uintptr_t buffer, size_t length,
+		      size_t *length_read)
```
Comment (danh-arm) [inline drivers/io/io_block.c]: Both the read and write functions could do with more comments or a big comment for each explaining what is going on.

File: include/drivers/io/io_block.h (hunk 2 above)
Comment (danh-arm) [inline include/drivers/io/io_block.h]: Later: Actually, it looks like `block_size` might need to be part of the "image spec"

File: drivers/io/io_block.c [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+				       count - skip);
+			}
+			left = left - (count - skip);
+		} else {
+			if (skip || padding)
+				count = block->ops.read(block->lba, buf->offset,
+							left);
+			else
+				count = block->ops.read(block->lba, buffer,
+							left);
```
Comment (danh-arm) [inline drivers/io/io_block.c]: So there are 4 different ways that the underlying `ops.read()` function is called. It's very difficult to understand what all the different cases are for without more comments.

File: include/drivers/io/io_block.h [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+ */
+
+#ifndef __IO_BLOCK_H__
+#define __IO_BLOCK_H__
+
+#include <io_storage.h>
+
+typedef struct io_block_ops {
+	size_t	(*read)(int lba, uintptr_t buf, size_t size);
+	size_t	(*write)(int lba, uintptr_t buf, size_t size);
```
Comment (danh-arm) [inline include/drivers/io/io_block.h]: `buf` arg in write function can be const.

File: include/drivers/emmc.h [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+	unsigned int		reserved_3:		2;
+	unsigned int		spec_vers:		4;
+	unsigned int		csd_structure:		2;
+} emmc_csd_t;
+
+size_t emmc_read_blocks(int lba, uintptr_t buf, size_t size);
+size_t emmc_write_blocks(int lba, uintptr_t buf, size_t size);
+size_t emmc_erase_blocks(int lba, size_t size);
+size_t emmc_rpmb_read_blocks(int lba, uintptr_t buf, size_t size);
+size_t emmc_rpmb_write_blocks(int lba, uintptr_t buf, size_t size);
```
Comment (danh-arm) [inline include/drivers/emmc.h]: buf arg in write function can be const

File: include/drivers/emmc.h [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+	unsigned int		tran_speed:		8;
+	unsigned int		nsac:			8;
+	unsigned int		taac:			8;
+	unsigned int		reserved_3:		2;
+	unsigned int		spec_vers:		4;
+	unsigned int		csd_structure:		2;
+} emmc_csd_t;
+
+size_t emmc_read_blocks(int lba, uintptr_t buf, size_t size);
+size_t emmc_write_blocks(int lba, uintptr_t buf, size_t size);
```
Comment (danh-arm) [inline include/drivers/emmc.h]: buf arg in write function can be const

File: drivers/emmc/emmc.c [hunk 14]
```diff
@@ -0,0 +1,347 @@
...
+	       (ops_ptr->read != 0) &&
+	       (ops_ptr->write != 0) &&
+	       (clk != 0) &&
+	       ((width == EMMC_BUS_WIDTH_1) ||
+		(width == EMMC_BUS_WIDTH_4) ||
+		(width == EMMC_BUS_WIDTH_8)));
+	ops = ops_ptr;
+
+	emmc_enumerate(clk, width);
+}
```
Comment (danh-arm) [inline drivers/emmc/emmc.c]: This driver looks good as far as I can see.

Out of interest, do you have any need for eMMC "boot operation" mode? It seems to me that it would be good to have a "cut down" eMMC driver that only supports reading data out of the eMMC boot partition. Such a driver would be more suitable for inclusion in ROM code than a full eMMC driver.

File: include/drivers/io/io_block.h (hunk 2 above)
Comment (hzhuang1) [inline include/drivers/io/io_block.h]: How about append a new helper in ops? like get_block_info()? Then I could get the block region and block size from this. Without region and block size, I can't calculate lba and file_pos.

File: drivers/emmc/emmc.c (hunk 14 above)
Comment (hzhuang1) [inline drivers/emmc/emmc.c]: I must access eMMC boot operation (RPMB). In HiKey recovery mode, I need to update everything including RPMB. Otherwise, I could make HiKey working again when it becomes brick.

File: include/drivers/io/io_block.h (hunk 4 above)
Comment (hzhuang1) [inline include/drivers/io/io_block.h]: Yes, I could remove it. Actually, the "rpmb_enable" field isn't used. I could declare different io_block_entity_t for boot area and user data area.

File: drivers/emmc/emmc.c (hunk 14 above)
Comment (danh-arm) [inline drivers/emmc/emmc.c]: Yes, if you want your recovery mode IO drivers in BL1, then this will need to include write and RPMB support. On some platforms, it may be possible to put this additional driver in a separate non-secure image, keeping BL1 minimal.

By the way, when I say eMMC boot operation mode, I don't mean RPMB; I mean reading from the boot partition before CMD1 is issued.

**Comment by hzhuang1** [comment]
Replaced by https://github.com/ARM-software/arm-trusted-firmware/pull/597
